【Hackathon 8th No.2】为 Paddle 新增 baddbmm API -part

Qin-sx opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->

User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Deprecations | Others ] -->

New features

### Description
<!-- Describe what you’ve done -->

https://github.com/PaddlePaddle/community/blob/master/hackathon/hackathon_8th/%E3%80%90Hackathon_8th%E3%80%91%E4%B8%AA%E4%BA%BA%E6%8C%91%E6%88%98%E8%B5%9B%E2%80%94%E6%A1%86%E6%9E%B6%E5%BC%80%E5%8F%91%E4%BB%BB%E5%8A%A1%E5%90%88%E9%9B%86.md#no2-%E4%B8%BA-paddle-%E6%96%B0%E5%A2%9E-baddbmm-api

新增加了float16和bfloat16的特例化

- RFC文档：https://github.com/PaddlePaddle/community/pull/1051
- 中文文档：https://github.com/PaddlePaddle/docs/pull/7041

Qin-sx committed:
added reset peak value initialization

	modified:   paddle/fluid/pybind/pybind.cc
	modified:   paddle/phi/core/memory/stats.cc
	modified:   paddle/phi/core/memory/stats.h
	modified:   python/paddle/device/cuda/__init__.py

Qin-sx committed:
added comments

	modified:   paddle/fluid/pybind/pybind.cc
	modified:   python/paddle/device/cuda/__init__.py

Qin-sx committed:
added cpp tests

	modified:   paddle/fluid/pybind/pybind.cc
	modified:   paddle/phi/core/memory/stats.cc
	modified:   paddle/phi/core/memory/stats.h
	modified:   test/cpp/fluid/memory/stats_test.cc

Qin-sx committed:
added python tests

	new file:   test/legacy_test/test_cuda_memory_stats.py
	new file:   test/legacy_test/test_cuda_reset_peak_memory_stats.py

Qin-sx committed:
added a python test for reset_max_memory_allocated

	new file:   test/legacy_test/test_cuda_reset_max_memory_allocated.py

Qin-sx committed:
formatted by pre-commit

	modified:   python/paddle/device/cuda/__init__.py
	modified:   test/legacy_test/test_cuda_memory_stats.py
	modified:   test/legacy_test/test_cuda_reset_max_memory_allocated.py
	modified:   test/legacy_test/test_cuda_reset_peak_memory_stats.py

Qin-sx committed:
formatted by pre-commit (clang-format)
	modified:   paddle/fluid/pybind/pybind.cc
	modified:   paddle/phi/core/memory/stats.cc
	modified:   paddle/phi/core/memory/stats.h
	modified:   test/cpp/fluid/memory/stats_test.cc

Qin-sx committed:
added reset max memory reserved function

	modified:   python/paddle/device/cuda/__init__.py
	modified:   test/legacy_test/test_cuda_reset_max_memory_allocated.py
	new file:   test/legacy_test/test_cuda_reset_max_memory_reserved.py

Qin-sx committed:
deleted memory stats and reset peak memory stats

	modified:   paddle/fluid/pybind/pybind.cc
	modified:   python/paddle/device/cuda/__init__.py
	deleted:    test/legacy_test/test_cuda_memory_stats.py
	deleted:    test/legacy_test/test_cuda_reset_peak_memory_stats.py

Qin-sx committed:
Merge branch 'develop' of https://github.com/Qin-sx/Paddle into develop

Qin-sx committed:
formatted baddbmm version

	modified:   paddle/phi/infermeta/ternary.cc
	modified:   paddle/phi/infermeta/ternary.h
	new file:   paddle/phi/kernels/baddbmm_grad_kernel.h
	new file:   paddle/phi/kernels/baddbmm_kernel.h
	new file:   paddle/phi/kernels/cpu/baddbmm_grad_kernel.cc
	new file:   paddle/phi/kernels/cpu/baddbmm_kernel.cc
	modified:   paddle/phi/kernels/funcs/blas/blas.h
	modified:   paddle/phi/kernels/funcs/blas/blas_impl.cu.h
	modified:   paddle/phi/kernels/funcs/blas/blas_impl.h
	new file:   paddle/phi/kernels/gpu/baddbmm_grad_kernel.cu
	new file:   paddle/phi/kernels/gpu/baddbmm_kernel.cu
	new file:   paddle/phi/kernels/impl/baddbmm_grad_kernel_impl.h
	new file:   paddle/phi/kernels/impl/baddbmm_kernel_impl.h
	modified:   paddle/phi/ops/yaml/backward.yaml
	modified:   paddle/phi/ops/yaml/ops.yaml
	modified:   python/paddle/__init__.py
	modified:   python/paddle/tensor/__init__.py
	modified:   python/paddle/tensor/math.py

Qin-sx committed:
added file for test

	modified:   paddle/fluid/pir/dialect/op_generator/decomp_interface_gen_op_list.py
	modified:   paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc
	modified:   paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.h
	modified:   paddle/fluid/primitive/decomp_rule/decomp_rule/composite.h
	modified:   paddle/phi/api/ext/tensor_compat.h
	modified:   paddle/phi/kernels/impl/baddbmm_grad_kernel_impl.h
	modified:   paddle/phi/ops/yaml/op_compat.yaml
	new file:   test/legacy_test/test_baddbmm_op.py

Qin-sx committed:
added bloat16 case

	modified:   ../paddle/phi/kernels/funcs/blas/blas_impl.cu.h
	modified:   ../paddle/phi/kernels/impl/baddbmm_kernel_impl.h

Qin-sx committed:
added InferSymbolicShapeInterface

	modified:   ../paddle/phi/ops/yaml/ops.yaml

Qin-sx committed:
added more tests

	modified:   ../test/legacy_test/test_baddbmm_op.py

Qin-sx committed:
tests overtime (20s), reduced size

	modified:   ../test/legacy_test/test_baddbmm_op.py

Qin-sx committed:
test overtime again, reduced size again

	modified:   ../test/legacy_test/test_baddbmm_op.py

Qin-sx committed:
reduce size for overtime again

	modified:   ../test/legacy_test/test_baddbmm_op.py

Qin-sx committed:
just one test for overtime

	modified:   ../test/legacy_test/test_baddbmm_op.py

Qin-sx committed:
divided tests in some files

	modified:   ../test/legacy_test/test_baddbmm_op.py
	new file:   ../test/legacy_test/test_baddbmm_op1.py
	new file:   ../test/legacy_test/test_baddbmm_op2.py
	new file:   ../test/legacy_test/test_baddbmm_op3.py

Qin-sx committed:
enable static

	modified:   ../test/legacy_test/test_baddbmm_op1.py

Qin-sx committed:
deleted baddbmm in tensor_compat.h

	modified:   ../paddle/phi/api/ext/tensor_compat.h

Qin-sx committed:
added float in hip

	modified:   ../paddle/phi/kernels/funcs/blas/blas_impl.hip.h

Qin-sx committed:
pre-commit

	modified:   ../paddle/phi/kernels/funcs/blas/blas_impl.hip.h

Qin-sx committed:
added more tests

	modified:   ../test/legacy_test/test_baddbmm_op3.py
	new file:   ../test/legacy_test/test_baddbmm_op4.py

Qin-sx committed:
Merge branch 'PaddlePaddle:develop' into develop

Qin-sx committed:
Merge branch 'PaddlePaddle:develop' into develop

DrownFish19 commented on paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc:
```diff
@@ -163,6 +163,67 @@ bool Addmm_OpInferSymbolicShape(pir::Operation *op,
   return AddmmOpInferSymbolicShape(op, infer_context);
 }
 
+bool BaddbmmOpInferSymbolicShape(
```
调整代码位置，函数名按照ABC排列

Qin-sx replied:
收到，已修改，谢谢

DrownFish19 commented on paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.h:
```diff
@@ -21,6 +21,8 @@ namespace paddle::dialect {
 OP_DECLARE_INFER_SYMBOLIC_SHAPE(Accuracy)
 OP_DECLARE_INFER_SYMBOLIC_SHAPE(Addmm)
 OP_DECLARE_INFER_SYMBOLIC_SHAPE(Addmm_)
+OP_DECLARE_INFER_SYMBOLIC_SHAPE(Baddbmm)
```
调整代码位置，函数名按照ABC排列

Qin-sx replied:
收到，已修改，谢谢

DrownFish19 commented on paddle/phi/kernels/baddbmm_grad_kernel.h:
```diff
@@ -0,0 +1,33 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
新增文件注意一下copyright 年份

Qin-sx replied:
收到，新增文件均已修改，谢谢

DrownFish19 commented on paddle/phi/kernels/funcs/blas/blas_impl.cu.h:
```diff
@@ -1183,6 +1183,152 @@ inline void Blas<phi::GPUContext>::GEMM(CBLAS_TRANSPOSE transA,
 }
 
+template <>
+template <typename T, typename U>
```
这里是否需要和.h一致 `template <typename T, typename U = T>`

DrownFish19 replied:
辛苦检查同级文件

Qin-sx replied:
这里在模板偏特化时，应该不能再设置默认的模板参数了。  
以下为报错提示：
```bash
/Paddle_2_GPU/Paddle/paddle/phi/kernels/funcs/blas/blas_impl.cu.h(1187): error: a default template argument cannot be specified on the declaration of a member of a class template outside of its class
```

DrownFish19 commented:
辛苦在PR描述中补充精度验证代码和精度误差

Qin-sx committed:
Merge branch 'PaddlePaddle:develop' into develop

Qin-sx committed:
modified func order and copyright year

	modified:   paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc
	modified:   paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.h
	modified:   paddle/phi/kernels/baddbmm_grad_kernel.h
	modified:   paddle/phi/kernels/baddbmm_kernel.h
	modified:   paddle/phi/kernels/cpu/baddbmm_grad_kernel.cc
	modified:   paddle/phi/kernels/cpu/baddbmm_kernel.cc
	modified:   paddle/phi/kernels/gpu/baddbmm_grad_kernel.cu
	modified:   paddle/phi/kernels/gpu/baddbmm_kernel.cu
	modified:   paddle/phi/kernels/impl/baddbmm_grad_kernel_impl.h
	modified:   paddle/phi/kernels/impl/baddbmm_kernel_impl.h
	modified:   test/legacy_test/test_baddbmm_op.py
	modified:   test/legacy_test/test_baddbmm_op1.py
	modified:   test/legacy_test/test_baddbmm_op2.py
	modified:   test/legacy_test/test_baddbmm_op3.py
	modified:   test/legacy_test/test_baddbmm_op4.py

Qin-sx committed:
updated tests

	modified:   test/legacy_test/CMakeLists.txt
	modified:   test/legacy_test/test_baddbmm_op.py
	new file:   test/legacy_test/test_baddbmm_op5.py

Qin-sx committed:
updated blas hip

	modified:   paddle/phi/kernels/funcs/blas/blas_impl.hip.h

Qin-sx commented:
> 辛苦在PR描述中补充精度验证代码和精度误差

验证代码添加在了[RFC文档](https://github.com/PaddlePaddle/community/pull/1051)中。 

以下为测试结果：
1. float32类型
```bash
max diff between paddle matmul and torch baddbmm is 0.0021514892578125
max diff between paddle baddbmm and torch baddbmm is 0.0
max diff between paddle baddbmm and paddle matmul is 0.0021514892578125
```

2. float16类型
```bash
max diff between paddle matmul and torch baddbmm is 0.25
max diff between paddle baddbmm and torch baddbmm is 0.0
max diff between paddle baddbmm and paddle matmul is 0.25
```

3. bfloat16类型
```bash
max diff between paddle matmul and torch baddbmm is 2.0
max diff between paddle baddbmm and torch baddbmm is 0.0
max diff between paddle baddbmm and paddle matmul is 2.0
```

Qin-sx commented:
- RFC文档：https://github.com/PaddlePaddle/community/pull/1051
- API文档：https://github.com/PaddlePaddle/docs/pull/7041

Qin-sx committed:
added broadcast

	modified:   paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc
	modified:   paddle/phi/infermeta/ternary.cc
	modified:   paddle/phi/kernels/impl/baddbmm_grad_kernel_impl.h
	modified:   paddle/phi/kernels/impl/baddbmm_kernel_impl.h
	modified:   python/paddle/tensor/math.py
	modified:   test/legacy_test/CMakeLists.txt
	new file:   test/legacy_test/test_baddbmm_op6.py

Qin-sx committed:
added more tests

	modified:   ../paddle/phi/infermeta/ternary.cc
	modified:   ../test/legacy_test/test_baddbmm_op3.py

Qin-sx committed:
added more tests for baddbmm_

	modified:   ../python/paddle/tensor/math.py
	new file:   ../test/legacy_test/test_baddbmm_op7.py

Qin-sx committed:
added more tests

	modified:   ../test/legacy_test/CMakeLists.txt
	modified:   ../test/legacy_test/test_baddbmm_op7.py

Qin-sx committed:
typo

	modified:   paddle/phi/kernels/funcs/blas/blas_impl.cu.h

Qin-sx committed:
Merge branch 'develop' into develop_baddbmm_format

Qin-sx commented:
关于PR-CI-Coverage没有覆盖到的两部分的相关issue  
https://github.com/PaddlePaddle/Paddle/issues/70848  
https://github.com/PaddlePaddle/Paddle/issues/70888

DrownFish19 commented on python/paddle/tensor/math.py:
```diff
@@ -2543,6 +2543,231 @@ def addmm_(
+                raise ValueError(
+                    f"If input's dimension[2] is not equal to y's dimension[2], input's dimension[2] must be 1. But received input's dimension[2] = {input_shape[2]}, y's dimension[2] = {y_shape[2]}"
+                )
+            # 上面的判断已经包含了这种情况
```
注释如不必要，可以删除

Qin-sx replied:
收到，已修改，谢谢

DrownFish19 commented on python/paddle/tensor/math.py:
```diff
@@ -2543,6 +2543,231 @@ def addmm_(
+                raise ValueError(
+                    f"If input's dimension[2] is not equal to y's dimension[2], input's dimension[2] must be 1. But received input's dimension[2] = {input_shape[2]}, y's dimension[2] = {y_shape[2]}"
+                )
+            # 上面的判断已经包含了这种情况
```
注释问题

Qin-sx replied:
收到，已修改，谢谢

Qin-sx committed:
deleted comments and 'print'

	modified:   ../paddle/phi/api/ext/tensor_compat.h
	modified:   ../python/paddle/tensor/math.py

DrownFish19 commented:
头文件覆盖率可豁免

DrownFish19 reviewed (approved):
LGTM

jeff41404 commented on python/paddle/tensor/math.py:
```diff
@@ -2543,6 +2543,215 @@ def addmm_(
+    name: str | None = None,
+) -> Tensor:
+    """
+    Inplace version of ``baddbmm`` API, the output Tensor will be inplaced with input ``x``.
```
the output Tensor will be inplaced with input ``input`` ? not ``x``?

Qin-sx replied:
I have modified it like  
```python
"""
Inplace version of ``baddbmm`` API, the output Tensor will be inplaced with input parameter ``x``.
"""
```
I referred to the comments of the `addmm_` function before.
```python
"""
Inplace version of ``addmm`` API, the output Tensor will be inplaced with input ``x``.
"""
  ```

Qin-sx replied:
I referred to the `addmm_` function before.
```python
"""
Inplace version of ``addmm`` API, the output Tensor will be inplaced with input ``x``.
"""
  ```
Should the output Tensor be inplace updated to input or x?

Qin-sx replied:
I referred to the `addmm_` function before.
```python
"""
Inplace version of ``addmm`` API, the output Tensor will be inplaced with input ``x``.
"""
  ```
Should the output Tensor be inplace updated to input or x?  
Or should I change the comments for the functions `addmm` and `baddmm` to 
```python
"""
, the output Tensor will be inplaced with input ``input ``.
"""
  ```

jeff41404 replied:
`inplace` means that the output needs to be written to a certain input(using the space of a certain input to save memory usage), which is determined by the yaml configuration. look to the configuration of [addmm]( https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/ops/yaml/ops.yaml#L149), when use addmm_, its output is written to a input parameter of `input`, the document of addmm_ needs to be modified correctly. so does baddbmm_ .

Qin-sx replied:
I have modified them. Thank you.

jeff41404 commented on test/legacy_test/CMakeLists.txt:
```diff
@@ -888,6 +888,15 @@ set_tests_properties(test_callback_wandb PROPERTIES TIMEOUT 60)
 set_tests_properties(test_jit_save_load PROPERTIES TIMEOUT 100)
 set_tests_properties(test_pool2d_op PROPERTIES TIMEOUT 120)
 
+set_tests_properties(test_baddbmm_op PROPERTIES TIMEOUT 120)
+set_tests_properties(test_baddbmm_op1 PROPERTIES TIMEOUT 120)
+set_tests_properties(test_baddbmm_op2 PROPERTIES TIMEOUT 120)
+set_tests_properties(test_baddbmm_op3 PROPERTIES TIMEOUT 120)
+set_tests_properties(test_baddbmm_op4 PROPERTIES TIMEOUT 120)
+set_tests_properties(test_baddbmm_op5 PROPERTIES TIMEOUT 120)
+set_tests_properties(test_baddbmm_op6 PROPERTIES TIMEOUT 120)
+set_tests_properties(test_baddbmm_op7 PROPERTIES TIMEOUT 120)
```
Why need 8 unit test files? Can we merge them all into test_baddbmm_op.py

Qin-sx replied:
I have merged the tests into a single file.   
Since I was not familiar with the timeout settings for tests before, I couldn't pass the tests in the cloud environment. Therefore, I split them into multiple files during debugging.

Qin-sx committed:
merged unit tests into one file

	modified:   test/legacy_test/test_baddbmm_op.py

Qin-sx committed:
deleted other test files

	modified:   test/legacy_test/CMakeLists.txt
	deleted:    test/legacy_test/test_baddbmm_op1.py
	deleted:    test/legacy_test/test_baddbmm_op2.py
	deleted:    test/legacy_test/test_baddbmm_op3.py
	deleted:    test/legacy_test/test_baddbmm_op4.py
	deleted:    test/legacy_test/test_baddbmm_op5.py
	deleted:    test/legacy_test/test_baddbmm_op6.py
	deleted:    test/legacy_test/test_baddbmm_op7.py

luotao1 commented on test/legacy_test/CMakeLists.txt:
```diff
@@ -888,6 +888,8 @@ set_tests_properties(test_callback_wandb PROPERTIES TIMEOUT 60)
 set_tests_properties(test_jit_save_load PROPERTIES TIMEOUT 100)
 set_tests_properties(test_pool2d_op PROPERTIES TIMEOUT 120)
 
+set_tests_properties(test_baddbmm_op PROPERTIES TIMEOUT 900)
```
<img width="710" alt="image" src="https://github.com/user-attachments/assets/c98a32d6-c929-4988-b92d-35c4870fe54d" />
从实际测试看，仅需21s，这里设置900s太多了吧。设置50s也够了。

Qin-sx replied:
但是感觉这个测试很不稳定，之前我拆分成多个测试时，有时候每个测试都会超过默认的20s。我猜测是不是多个测试或者多个任务，同时在同一个服务器上运行，会互相影响。要不我先改为500s可以吗？

luotao1 replied:
> 要不我先改为500s可以吗 

可以再重新构建下 coverage 流水线，看新一次测试跑多久，然后看改成多少秒。

Qin-sx replied:
收到，那我先改为100秒吧。因为之前我在PR-CI-Coverage中也遇到过其他测试超时的情况，需要手动重新构建。我还是感觉测试环境偶尔会不稳定，导致测试耗时增加。

Qin-sx committed:
reduced timeout

	modified:   test/legacy_test/CMakeLists.txt

Qin-sx committed:
modified comments for inplace functions

	modified:   python/paddle/tensor/math.py

jeff41404 reviewed (approved):
LGTM

sunzhongkai588 reviewed (approved):
LGTM for docs。 

文档渲染 CI 出了些问题，等合入后在官网再看看效果
